libraries/common/stream.effekt

```diff
@@ -0,0 +1,234 @@
+module stream
+
+import bytes
+
+interface Emit[A] {
+  def emit(value: A): Unit
+}
+
+def for[A, R] { stream: () => R / Emit[A] } { action: A => Unit }: R =
+  try {
+    stream()
+  } with Emit[A] {
+    def emit(value) = resume(action(value))
+  }
+
+// TODO what if the number is negative? undefined behavior?
+def repeat[A](number: Int) { action: () => A }: Unit / Emit[A] =
+  if (number > 0) {
+    do emit(action())
+    repeat(number - 1) {action}
+  }
+
+
+def each[A](list: List[A]): Unit / Emit[A] =
+  list match {
+    case Nil() => ()
+    case Cons(head, tail) =>
+      do emit(head)
+      each(tail)
+  }
+
+def each[A](array: Array[A]): Unit / Emit[A] = {
+  val n = array.size
+  def go(i: Int): Unit = {
+    if (i < n) {
+      do emit(array.unsafeGet(i))
+      go(i + 1)
+    }
+  }
+  go(0)
+}
+
+def each(bytes: Bytes): Unit / Emit[Byte] = {
+  val n = bytes.size
+  def go(i: Int): Unit = {
+    if (i < n) {
+      do emit(bytes.read(i))
+      go(i + 1)
+    }
+  }
+  go(0)
+}
+
+// not option
+// not dequeue
+// not queue
+// not result
+// not seq
+
+// def decodeUtf8[R] { stream: () => R / Emit[Byte] }: R / Emit[Char] = <>
+// def encodeUtf8[R] { stream: () => R / Emit[Char] }: R / Emit[Byte] = <>
+// def each(string: String): Unit / Emit[Char] = decodeUtf8 { string.asBytes.each }
+
+
+interface Done {
+  def done(): Nothing
+}
+
+def noneOnDone[R] { program: () => R / Done }: Option[R] =
+  try {
+    Some(program())
+  } with Done {
+    def done() = None()
+  }
+
+
+record Indexed[A](index: Int, value: A)
+
+// In Effekt lower bounds are inclusive and upper bounds are exclusive
+
+def range(lower: Int, upper: Int): Unit / Emit[Int] =
+  if (lower < upper) {
+    do emit(lower)
+    range(lower + 1, upper)
+  }
+
+def rangeTo(upper: Int): Unit / Emit[Int] =
+  range(0, upper)
+
+def rangeFrom(lower: Int): Unit / Emit[Int] = {
+  do emit(lower)
+  rangeFrom(lower + 1)
+}
+
+
+def index[A, R] { stream: () => R / Emit[A] }: R / Emit[Indexed[A]] = {
+  var i = 0;
+  try {
+    stream()
+  } with Emit[A] {
+    def emit(v) = {
+      val c = i;
+      i = c + 1;
+      resume(do emit(Indexed(c, v)))
+    }
+  }
+}
+
+def ever[A] { action: () => A }: Nothing / Emit[A] = {
+  def go(): Nothing = {
+    do emit(action())
+    go()
+  }
+  go()
+}
+
+def limit[A, R](number: Int) { stream: () => R / Emit[A] }: R / { Emit[A], Done } = {
+  var i = 0;
+  try {
+    stream()
+  } with Emit[A] {
+    def emit(v) = {
+      val c = i;
+      if (c < number) {
+        i = c + 1;
+        resume(do emit(v))
+      } else {
+        do done()
+      }
+    }
+  }
+}
```

Comment:
This is the same as:

```suggestion
def limit[A, R](number: Int) { stream: () => R / Emit[A] }: R / { Emit[A], Done } = {
  var i = 0;
  for { stream() } {
    if (c < number) { i = c + 1; do emit(v) }
    else { do done() }
  }
}
```

isn't it?

Reply:
Also `limit` seems like it would limit the emitted values, not their number. Why not the standard `take`?